internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,313 @@
+
+// Set sets the value for key k in the MemBuffer.
+func (p *PipelinedMemDB) Set(key, value []byte) error {
+	return p.memBuffer.Set(key, value)
+}
+
+// SetWithFlags sets the value for key k in the MemBuffer with flags.
+func (p *PipelinedMemDB) SetWithFlags(key, value []byte, ops ...kv.FlagsOp) error {
+	return p.memBuffer.SetWithFlags(key, value, ops...)
+}
+
+// Delete deletes the key k in the MemBuffer.
+func (p *PipelinedMemDB) Delete(key []byte) error {
+	return p.memBuffer.Delete(key)
+}
+
+// DeleteWithFlags deletes the key k in the MemBuffer with flags.
+func (p *PipelinedMemDB) DeleteWithFlags(key []byte, ops ...kv.FlagsOp) error {
+	return p.memBuffer.DeleteWithFlags(key, ops...)
+}
+
+// Flush is called during execution of a transaction, it does flush when there are enough keys and the ongoing flushingMemBuffer is done.
+// When the mutable memdb is too large, it blocks until the ongoing flush is done.
+func (p *PipelinedMemDB) Flush(force bool) (bool, error) {
+	if p.flushFunc == nil {
+		return false, errors.New("flushFunc is not provided")
+	}
+	if !force && !p.needFlush() {
+		return false, nil
+	}
+	if p.flushingMemBuffer != nil {
+		if err := <-p.errCh; err != nil {
+			return false, err
+		}
+	}
+	p.onFlushing.Store(true)
+	p.flushingMemBuffer = p.memBuffer
+	p.len += p.flushingMemBuffer.Len()
+	p.size += p.flushingMemBuffer.Size()
+	p.memBuffer = newMemDB()
+	p.generation++
+	go func(generation uint64) {
+		util.EvalFailpoint("beforePipelinedFlush")
+		metrics.TiKVPipelinedFlushLenHistogram.Observe(float64(p.flushingMemBuffer.Len()))
+		metrics.TiKVPipelinedFlushSizeHistogram.Observe(float64(p.flushingMemBuffer.Size()))
+		err := p.flushFunc(generation, p.flushingMemBuffer)
+		p.onFlushing.Store(false)
+		// Send the error to errCh after onFlushing status is set to false.
+		// this guarantees the onFlushing.Store(true) in another goroutine's MayFlush happens after onFlushing.Store(false) in this function.
+		p.errCh <- err
+	}(p.generation)
+	return true, nil
+}
+
+func (p *PipelinedMemDB) needFlush() bool {
+	size := p.memBuffer.Size()
+	// size < MinFlushSize, do not flush.
+	// MinFlushSize <= size < MaxFlushSize && keys < MinFlushKeys, do not flush.
+	// MinFlushSize <= size < MaxFlushSize && keys >= MinFlushKeys, flush.
+	// size >= MaxFlushSize, flush.
+	if size < MinFlushSize || (p.memBuffer.Len() < MinFlushKeys && size < MaxFlushSize) {
+		return false
+	}
+	if p.onFlushing.Load() && size < MaxFlushSize {
+		return false
+	}
+	return true
+}
+
+// FlushWait will wait for all flushing tasks are done and return the error if there is a failure.
+func (p *PipelinedMemDB) FlushWait() error {
+	if p.flushingMemBuffer != nil {
+		err := <-p.errCh
+		// cleanup the flushingMemBuffer so the next call of FlushWait will not wait for the error channel.
+		p.flushingMemBuffer = nil
+		return err
+	}
+	return nil
+}
+
+// Iter implements the Retriever interface.
+func (p *PipelinedMemDB) Iter([]byte, []byte) (Iterator, error) {
```

Comment:
Do we need to return an emtpy iter here? What will happen during a UnionIter?

Reply:
An empty iter can lead to wrong resultset. Image the following usage:

```
begin()
put(k, 1)
flush()
iter() // this equals to snapshot iter if an empty returned from PipelinedMemDB.Iter, put(k, 1) is missing
```

Reply:
`Iter` and `IterReverse` return error directly, and I just added an error iterator for `SnapshotIter` and `SnapshotIterReverse`, they should never be called for pipelined memdb.

In TiDB, those iterators are only used in unionscan, which never happens in autocommit statement.

https://github.com/tikv/client-go/blob/d480f284026dacd3d50e72e826fa2f1ace6c0fd3/internal/unionstore/pipelined_memdb.go#L261-L279